#BRCA STUDY

## 0. General

(10X)
Carriers: 23, 28, 35, 43R, 50, 55, 63, 65  
Controls: 16, 30, 33, 49, 51, 54, 57, 60, 70, 103, 106, 110
 
(Illumina)
Carriers: 7, 18, 21, 22, 35, 43L/R, 50, 52, 55, 63, 65, 69, 102

Responders: 16, 33, 70, 103, 106, 110  
Non-Responders: 30, 49, 51, 54, 57, 60

`T103_longranger/phased_possorted_bam.bam`

```
825465055 + 0 in total (QC-passed reads + QC-failed reads)  
56534097 + 0 secondary  
0 + 0 supplementary  
39296016 + 0 duplicates
791794254 + 0 mapped (95.92% : N/A)
768930958 + 0 paired in sequencing
384463443 + 0 read1
384467515 + 0 read2
587973853 + 0 properly paired (76.47% : N/A)
717955841 + 0 with itself and mate mapped
17304316 + 0 singletons (2.25% : N/A)
39222124 + 0 with mate mapped to a different chr
31798145 + 0 with mate mapped to a different chr (mapQ>=5)
```
517497020/791794254 = 65% reads phased

`T106_longranger/phased_possorted_bam.bam`

```
828066282 + 0 in total (QC-passed reads + QC-failed reads)
54167252 + 0 secondary
0 + 0 supplementary
89302766 + 0 duplicates
811572171 + 0 mapped (98.01% : N/A)
773899030 + 0 paired in sequencing
386949515 + 0 read1
386949515 + 0 read2
638677867 + 0 properly paired (82.53% : N/A)
747593607 + 0 with itself and mate mapped
9811312 + 0 singletons (1.27% : N/A)
36853408 + 0 with mate mapped to a different chr
30374622 + 0 with mate mapped to a different chr (mapQ>=5)
```
513066179/811572171 = 63% reads phased

`T110_longranger/phased_possorted_bam.bam`

```
816981573 + 0 in total (QC-passed reads + QC-failed reads)
56193476 + 0 secondary
0 + 0 supplementary
56732203 + 0 duplicates
798407601 + 0 mapped (97.73% : N/A)
760788097 + 0 paired in sequencing
380392356 + 0 read1
380395741 + 0 read2
629107559 + 0 properly paired (82.69% : N/A)
731328941 + 0 with itself and mate mapped
10885184 + 0 singletons (1.43% : N/A)
30807558 + 0 with mate mapped to a different chr
24861303 + 0 with mate mapped to a different chr (mapQ>=5)
```
498211460/798407601 = 62% reads phased

`T16_longranger/phased_possorted_bam.bam`

```728758280 + 0 in total (QC-passed reads + QC-failed reads)
43949486 + 0 secondary
0 + 0 supplementary
58968054 + 0 duplicates
695791375 + 0 mapped (95.48% : N/A)
684808794 + 0 paired in sequencing
342404397 + 0 read1
342404397 + 0 read2
531281707 + 0 properly paired (77.58% : N/A)
635592774 + 0 with itself and mate mapped
16249115 + 0 singletons (2.37% : N/A)
29238378 + 0 with mate mapped to a different chr
23843119 + 0 with mate mapped to a different chr (mapQ>=5)
```
355443450/695791375 = 51% reads phased

`T23_longranger/phased_possorted_bam.bam`

```
831978738 + 0 in total (QC-passed reads + QC-failed reads)
61222466 + 0 secondary
0 + 0 supplementary
46160737 + 0 duplicates
790076297 + 0 mapped (94.96% : N/A)
770756272 + 0 paired in sequencing
385378136 + 0 read1
385378136 + 0 read2
590636418 + 0 properly paired (76.63% : N/A)
716213666 + 0 with itself and mate mapped
12640165 + 0 singletons (1.64% : N/A)
36791840 + 0 with mate mapped to a different chr
29060372 + 0 with mate mapped to a different chr (mapQ>=5)
```
472522829/790076297 = 60% reads phased

`T28_longranger/phased_possorted_bam.bam`

```
896039560 + 0 in total (QC-passed reads + QC-failed reads)
65269000 + 0 secondary
0 + 0 supplementary
70314985 + 0 duplicates
851926986 + 0 mapped (95.08% : N/A)
830770560 + 0 paired in sequencing
415385280 + 0 read1
415385280 + 0 read2
630831094 + 0 properly paired (75.93% : N/A)
770967399 + 0 with itself and mate mapped
15690587 + 0 singletons (1.89% : N/A)
37428886 + 0 with mate mapped to a different chr
29346878 + 0 with mate mapped to a different chr (mapQ>=5)
```
502159853/851926986 = 59% reads phased

`T30_longranger/phased_possorted_bam.bam`

```
822139467 + 0 in total (QC-passed reads + QC-failed reads)
58217461 + 0 secondary
0 + 0 supplementary
172202378 + 0 duplicates
809303502 + 0 mapped (98.44% : N/A)
763922006 + 0 paired in sequencing
381961003 + 0 read1
381961003 + 0 read2
649344917 + 0 properly paired (85.00% : N/A)
744155474 + 0 with itself and mate mapped
6930567 + 0 singletons (0.91% : N/A)
25838076 + 0 with mate mapped to a different chr
20684380 + 0 with mate mapped to a different chr (mapQ>=5)
```
480374335/809303502 = 59% reads phased

`T33_longranger/phased_possorted_bam.bam`

```
805964704 + 0 in total (QC-passed reads + QC-failed reads)
53576232 + 0 secondary
0 + 0 supplementary
32982723 + 0 duplicates
771740612 + 0 mapped (95.75% : N/A)
752388472 + 0 paired in sequencing
376193061 + 0 read1
376195411 + 0 read2
577388229 + 0 properly paired (76.74% : N/A)
699912864 + 0 with itself and mate mapped
18251516 + 0 singletons (2.43% : N/A)
37950803 + 0 with mate mapped to a different chr
31125020 + 0 with mate mapped to a different chr (mapQ>=5)
```
431606195/771740612 = 56% reads phased

`T35_longranger/phased_possorted_bam.bam`

```
908607047 + 0 in total (QC-passed reads + QC-failed reads)
62764903 + 0 secondary
0 + 0 supplementary
86278140 + 0 duplicates
862649577 + 0 mapped (94.94% : N/A)
845842144 + 0 paired in sequencing
422918933 + 0 read1
422923211 + 0 read2
640381406 + 0 properly paired (75.71% : N/A)
783071864 + 0 with itself and mate mapped
16812810 + 0 singletons (1.99% : N/A)
42930492 + 0 with mate mapped to a different chr
34187339 + 0 with mate mapped to a different chr (mapQ>=5)
```
440895208/862649577 = 51% reads phased

`T43R_longranger/phased_possorted_bam.bam`

```
762188335 + 0 in total (QC-passed reads + QC-failed reads)
55975372 + 0 secondary
0 + 0 supplementary
37639941 + 0 duplicates
718494947 + 0 mapped (94.27% : N/A)
706212963 + 0 paired in sequencing
353104931 + 0 read1
353108032 + 0 read2
532055619 + 0 properly paired (75.34% : N/A)
648058015 + 0 with itself and mate mapped
14461560 + 0 singletons (2.05% : N/A)
36460138 + 0 with mate mapped to a different chr
28341849 + 0 with mate mapped to a different chr (mapQ>=5)
```
389612652/718494947 = 54% reads phased

`T49_longranger/phased_possorted_bam.bam`

```
850319782 + 0 in total (QC-passed reads + QC-failed reads)
59468630 + 0 secondary
0 + 0 supplementary
39611498 + 0 duplicates
816501962 + 0 mapped (96.02% : N/A)
790851152 + 0 paired in sequencing
395424909 + 0 read1
395426243 + 0 read2
604333463 + 0 properly paired (76.42% : N/A)
739793633 + 0 with itself and mate mapped
17239699 + 0 singletons (2.18% : N/A)
40432650 + 0 with mate mapped to a different chr
33011838 + 0 with mate mapped to a different chr (mapQ>=5)
```
432118551/850319782 = 51% reads phased

`T50_longranger/phased_possorted_bam.bam`

```
874554775 + 0 in total (QC-passed reads + QC-failed reads)
64209373 + 0 secondary
0 + 0 supplementary
75801925 + 0 duplicates
810772100 + 0 mapped (92.71% : N/A)
810345402 + 0 paired in sequencing
405170554 + 0 read1
405174848 + 0 read2
588328035 + 0 properly paired (72.60% : N/A)
728960331 + 0 with itself and mate mapped
17602396 + 0 singletons (2.17% : N/A)
40844198 + 0 with mate mapped to a different chr
32262367 + 0 with mate mapped to a different chr (mapQ>=5)
```
467264546/810772100 = 57% reads phased

`T57_longranger/phased_possorted_bam.bam`

```
849186470 + 0 in total (QC-passed reads + QC-failed reads)
57971824 + 0 secondary
0 + 0 supplementary
44575784 + 0 duplicates
817209674 + 0 mapped (96.23% : N/A)
791214646 + 0 paired in sequencing
395607323 + 0 read1
395607323 + 0 read2
606070612 + 0 properly paired (76.60% : N/A)
742564860 + 0 with itself and mate mapped
16672990 + 0 singletons (2.11% : N/A)
40376824 + 0 with mate mapped to a different chr
32400274 + 0 with mate mapped to a different chr (mapQ>=5)
```
459235558/817209674 = 56% reads phased

`T63_longranger/phased_possorted_bam.bam`

```
739709736 + 0 in total (QC-passed reads + QC-failed reads)
51703907 + 0 secondary
0 + 0 supplementary
47026417 + 0 duplicates
699350286 + 0 mapped (94.54% : N/A)
688005829 + 0 paired in sequencing
344001558 + 0 read1
344004271 + 0 read2
506283825 + 0 properly paired (73.59% : N/A)
631994526 + 0 with itself and mate mapped
15651853 + 0 singletons (2.27% : N/A)
35036257 + 0 with mate mapped to a different chr
27408814 + 0 with mate mapped to a different chr (mapQ>=5)
```
392134698/699350286 = 56% reads phased

`T65_longranger/phased_possorted_bam.bam`

```
932401918 + 0 in total (QC-passed reads + QC-failed reads)
73765744 + 0 secondary
0 + 0 supplementary
111744163 + 0 duplicates
859298647 + 0 mapped (92.16% : N/A)
858636174 + 0 paired in sequencing
429318087 + 0 read1
429318087 + 0 read2
634485637 + 0 properly paired (73.89% : N/A)
771119294 + 0 with itself and mate mapped
14413609 + 0 singletons (1.68% : N/A)
38022884 + 0 with mate mapped to a different chr
30995084 + 0 with mate mapped to a different chr (mapQ>=5)
```
484826799/859298647 = 56% reads phased

`T70_longranger/phased_possorted_bam.bam`

```
812962705 + 0 in total (QC-passed reads + QC-failed reads)
54680233 + 0 secondary
0 + 0 supplementary
34737648 + 0 duplicates
780287080 + 0 mapped (95.98% : N/A)
758282472 + 0 paired in sequencing
379141236 + 0 read1
379141236 + 0 read2
577008358 + 0 properly paired (76.09% : N/A)
708799403 + 0 with itself and mate mapped
16807444 + 0 singletons (2.22% : N/A)
39048984 + 0 with mate mapped to a different chr
31470603 + 0 with mate mapped to a different chr (mapQ>=5)
```
432506594/780287080 = 55% reads phased

`T51_longranger/phased_possorted_bam.bam`

```
800861620 + 0 in total (QC-passed reads + QC-failed reads)
57306252 + 0 secondary
0 + 0 supplementary
81148488 + 0 duplicates
782453523 + 0 mapped (97.70% : N/A)
743555368 + 0 paired in sequencing
371776878 + 0 read1
371778490 + 0 read2
613080563 + 0 properly paired (82.45% : N/A)
715189959 + 0 with itself and mate mapped
9957312 + 0 singletons (1.34% : N/A)
30158548 + 0 with mate mapped to a different chr
24403467 + 0 with mate mapped to a different chr (mapQ>=5)
T54_longranger/phased_possorted_bam.bam
821873760 + 0 in total (QC-passed reads + QC-failed reads)
55944010 + 0 secondary
```
529925881/782453523 = 68% reads phased

`T54_longranger/phased_possorted_bam.bam`

```
821873760 + 0 in total (QC-passed reads + QC-failed reads)
55944010 + 0 secondary
0 + 0 supplementary
55633727 + 0 duplicates
788122501 + 0 mapped (95.89% : N/A)
765929750 + 0 paired in sequencing
382964875 + 0 read1
382964875 + 0 read2
586716726 + 0 properly paired (76.60% : N/A)
716096565 + 0 with itself and mate mapped
16081926 + 0 singletons (2.10% : N/A)
38753726 + 0 with mate mapped to a different chr
31117240 + 0 with mate mapped to a different chr (mapQ>=5)
```
401631093/788122501 = 51% reads phased

`T55_longranger/phased_possorted_bam.bam`

```
904029766 + 0 in total (QC-passed reads + QC-failed reads)
61716032 + 0 secondary
0 + 0 supplementary
70847729 + 0 duplicates
856548001 + 0 mapped (94.75% : N/A)
842313734 + 0 paired in sequencing
421155738 + 0 read1
421157996 + 0 read2
635074974 + 0 properly paired (75.40% : N/A)
777441984 + 0 with itself and mate mapped
17389985 + 0 singletons (2.06% : N/A)
40091536 + 0 with mate mapped to a different chr
31683816 + 0 with mate mapped to a different chr (mapQ>=5)
```
495167782/856548001 = 58% reads phased


`T60_longranger/phased_possorted_bam.bam`

```
800669575 + 0 in total (QC-passed reads + QC-failed reads)
58483847 + 0 secondary
0 + 0 supplementary
35675376 + 0 duplicates
769335804 + 0 mapped (96.09% : N/A)
742185728 + 0 paired in sequencing
371092864 + 0 read1
371092864 + 0 read2
558072672 + 0 properly paired (75.19% : N/A)
692396243 + 0 with itself and mate mapped
18455714 + 0 singletons (2.49% : N/A)
43243626 + 0 with mate mapped to a different chr
35462656 + 0 with mate mapped to a different chr (mapQ>=5)
```

418241004/769335804 = 54% reads phased

## Non-10X data

`B102.bam`

```
708160256 + 0 in total (QC-passed reads + QC-failed reads)
4310086 + 0 secondary
0 + 0 supplementary
92629084 + 0 duplicates
704506823 + 0 mapped (99.48% : N/A)
703850170 + 0 paired in sequencing
351925085 + 0 read1
351925085 + 0 read2
684097168 + 0 properly paired (97.19% : N/A)
698245578 + 0 with itself and mate mapped
1951159 + 0 singletons (0.28% : N/A)
9567886 + 0 with mate mapped to a different chr
4903580 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B18.bam`

```
880693224 + 0 in total (QC-passed reads + QC-failed reads)
6050426 + 0 secondary
0 + 0 supplementary
110332418 + 0 duplicates
876528329 + 0 mapped (99.53% : N/A)
874642798 + 0 paired in sequencing
437321399 + 0 read1
437321399 + 0 read2
849219874 + 0 properly paired (97.09% : N/A)
868028382 + 0 with itself and mate mapped
2449521 + 0 singletons (0.28% : N/A)
12181590 + 0 with mate mapped to a different chr
5648612 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B21.bam`

```
728485433 + 0 in total (QC-passed reads + QC-failed reads)
4666325 + 0 secondary
0 + 0 supplementary
101630043 + 0 duplicates
724577456 + 0 mapped (99.46% : N/A)
723819108 + 0 paired in sequencing
361909554 + 0 read1
361909554 + 0 read2
702325352 + 0 properly paired (97.03% : N/A)
717838906 + 0 with itself and mate mapped
2072225 + 0 singletons (0.29% : N/A)
10741194 + 0 with mate mapped to a different chr
5972428 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B22.bam`

```
691709257 + 0 in total (QC-passed reads + QC-failed reads)
4326535 + 0 secondary
0 + 0 supplementary
75301457 + 0 duplicates
686851596 + 0 mapped (99.30% : N/A)
687382722 + 0 paired in sequencing
343691361 + 0 read1
343691361 + 0 read2
667521980 + 0 properly paired (97.11% : N/A)
680758132 + 0 with itself and mate mapped
1766929 + 0 singletons (0.26% : N/A)
9126386 + 0 with mate mapped to a different chr
4721286 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B35.bam`

```
877574264 + 0 in total (QC-passed reads + QC-failed reads)
5296774 + 0 secondary
0 + 0 supplementary
99477645 + 0 duplicates
872996485 + 0 mapped (99.48% : N/A)
872277490 + 0 paired in sequencing
436138745 + 0 read1
436138745 + 0 read2
848158726 + 0 properly paired (97.23% : N/A)
864873278 + 0 with itself and mate mapped
2826433 + 0 singletons (0.32% : N/A)
11091372 + 0 with mate mapped to a different chr
5369038 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B43.bam`

```
827718694 + 0 in total (QC-passed reads + QC-failed reads)
5393796 + 0 secondary
0 + 0 supplementary
89633508 + 0 duplicates
820840919 + 0 mapped (99.17% : N/A)
822324898 + 0 paired in sequencing
411162449 + 0 read1
411162449 + 0 read2
795908904 + 0 properly paired (96.79% : N/A)
812725644 + 0 with itself and mate mapped
2721479 + 0 singletons (0.33% : N/A)
11567870 + 0 with mate mapped to a different chr
5796860 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B50.bam`

```
719852169 + 0 in total (QC-passed reads + QC-failed reads)
4553047 + 0 secondary
0 + 0 supplementary
91945770 + 0 duplicates
716157099 + 0 mapped (99.49% : N/A)
715299122 + 0 paired in sequencing
357649561 + 0 read1
357649561 + 0 read2
694988176 + 0 properly paired (97.16% : N/A)
709919584 + 0 with itself and mate mapped
1684468 + 0 singletons (0.24% : N/A)
9911340 + 0 with mate mapped to a different chr
5020517 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B52.bam`

```
830817518 + 0 in total (QC-passed reads + QC-failed reads)
5517948 + 0 secondary
0 + 0 supplementary
105889989 + 0 duplicates
827046846 + 0 mapped (99.55% : N/A)
825299570 + 0 paired in sequencing
412649785 + 0 read1
412649785 + 0 read2
801743060 + 0 properly paired (97.15% : N/A)
819327930 + 0 with itself and mate mapped
2200968 + 0 singletons (0.27% : N/A)
11739224 + 0 with mate mapped to a different chr
5782132 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B55.bam`

```
784803723 + 0 in total (QC-passed reads + QC-failed reads)
4904095 + 0 secondary
0 + 0 supplementary
105455597 + 0 duplicates
780661805 + 0 mapped (99.47% : N/A)
779899628 + 0 paired in sequencing
389949814 + 0 read1
389949814 + 0 read2
758706560 + 0 properly paired (97.28% : N/A)
774066692 + 0 with itself and mate mapped
1691018 + 0 singletons (0.22% : N/A)
10254578 + 0 with mate mapped to a different chr
5220746 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B63.bam`

```
823186863 + 0 in total (QC-passed reads + QC-failed reads)
4960571 + 0 secondary
0 + 0 supplementary
109853753 + 0 duplicates
819482274 + 0 mapped (99.55% : N/A)
818226292 + 0 paired in sequencing
409113146 + 0 read1
409113146 + 0 read2
796371650 + 0 properly paired (97.33% : N/A)
812441976 + 0 with itself and mate mapped
2079727 + 0 singletons (0.25% : N/A)
10720964 + 0 with mate mapped to a different chr
5353886 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B65.bam`

```
762232296 + 0 in total (QC-passed reads + QC-failed reads)
4638374 + 0 secondary
0 + 0 supplementary
100164915 + 0 duplicates
758594495 + 0 mapped (99.52% : N/A)
757593922 + 0 paired in sequencing
378796961 + 0 read1
378796961 + 0 read2
736850352 + 0 properly paired (97.26% : N/A)
751982728 + 0 with itself and mate mapped
1973393 + 0 singletons (0.26% : N/A)
10107910 + 0 with mate mapped to a different chr
5098913 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B69.bam`

```
896117408 + 0 in total (QC-passed reads + QC-failed reads)
5665878 + 0 secondary
0 + 0 supplementary
116447078 + 0 duplicates
891952519 + 0 mapped (99.54% : N/A)
890451530 + 0 paired in sequencing
445225765 + 0 read1
445225765 + 0 read2
866201124 + 0 properly paired (97.28% : N/A)
884054820 + 0 with itself and mate mapped
2231821 + 0 singletons (0.25% : N/A)
11897682 + 0 with mate mapped to a different chr
5728615 + 0 with mate mapped to a different chr (mapQ>=5)
```

`B7.bam`

```
821434414 + 0 in total (QC-passed reads + QC-failed reads)
5305456 + 0 secondary
0 + 0 supplementary
106471049 + 0 duplicates
817267299 + 0 mapped (99.49% : N/A)
816128958 + 0 paired in sequencing
408064479 + 0 read1
408064479 + 0 read2
791951222 + 0 properly paired (97.04% : N/A)
809737354 + 0 with itself and mate mapped
2224489 + 0 singletons (0.27% : N/A)
12325804 + 0 with mate mapped to a different chr
6826932 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T102.bam`

```
703930714 + 0 in total (QC-passed reads + QC-failed reads)
4061950 + 0 secondary
0 + 0 supplementary
73521299 + 0 duplicates
700151349 + 0 mapped (99.46% : N/A)
699868764 + 0 paired in sequencing
349934382 + 0 read1
349934382 + 0 read2
679690126 + 0 properly paired (97.12% : N/A)
693915888 + 0 with itself and mate mapped
2173511 + 0 singletons (0.31% : N/A)
9196584 + 0 with mate mapped to a different chr
4566547 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T18.bam`

```
668180738 + 0 in total (QC-passed reads + QC-failed reads)
4463956 + 0 secondary
0 + 0 supplementary
72320920 + 0 duplicates
662451606 + 0 mapped (99.14% : N/A)
663716782 + 0 paired in sequencing
331858391 + 0 read1
331858391 + 0 read2
641738906 + 0 properly paired (96.69% : N/A)
655923418 + 0 with itself and mate mapped
2064232 + 0 singletons (0.31% : N/A)
9147432 + 0 with mate mapped to a different chr
4287851 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T21.bam`

```
722089448 + 0 in total (QC-passed reads + QC-failed reads)
4168986 + 0 secondary
0 + 0 supplementary
80351909 + 0 duplicates
715182903 + 0 mapped (99.04% : N/A)
717920462 + 0 paired in sequencing
358960231 + 0 read1
358960231 + 0 read2
694396526 + 0 properly paired (96.72% : N/A)
708570454 + 0 with itself and mate mapped
2443463 + 0 singletons (0.34% : N/A)
9099656 + 0 with mate mapped to a different chr
4617015 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T22.bam`

```
848873054 + 0 in total (QC-passed reads + QC-failed reads)
5056416 + 0 secondary
0 + 0 supplementary
100459945 + 0 duplicates
845224336 + 0 mapped (99.57% : N/A)
843816638 + 0 paired in sequencing
421908319 + 0 read1
421908319 + 0 read2
823273922 + 0 properly paired (97.57% : N/A)
838788980 + 0 with itself and mate mapped
1378940 + 0 singletons (0.16% : N/A)
9451502 + 0 with mate mapped to a different chr
4217836 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T35.bam`

```
685904191 + 0 in total (QC-passed reads + QC-failed reads)
4155345 + 0 secondary
0 + 0 supplementary
81034585 + 0 duplicates
681495903 + 0 mapped (99.36% : N/A)
681748846 + 0 paired in sequencing
340874423 + 0 read1
340874423 + 0 read2
662493836 + 0 properly paired (97.18% : N/A)
675315496 + 0 with itself and mate mapped
2025062 + 0 singletons (0.30% : N/A)
8274324 + 0 with mate mapped to a different chr
3854308 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T43L.bam`

```
775569630 + 0 in total (QC-passed reads + QC-failed reads)
4426554 + 0 secondary
0 + 0 supplementary
75894523 + 0 duplicates
771003367 + 0 mapped (99.41% : N/A)
771143076 + 0 paired in sequencing
385571538 + 0 read1
385571538 + 0 read2
749902702 + 0 properly paired (97.25% : N/A)
763915252 + 0 with itself and mate mapped
2661561 + 0 singletons (0.35% : N/A)
9275980 + 0 with mate mapped to a different chr
4559934 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T43R.bam`

```
950394570 + 0 in total (QC-passed reads + QC-failed reads)
5938196 + 0 secondary
0 + 0 supplementary
114378395 + 0 duplicates
945748604 + 0 mapped (99.51% : N/A)
944456374 + 0 paired in sequencing
472228187 + 0 read1
472228187 + 0 read2
919187614 + 0 properly paired (97.32% : N/A)
937315746 + 0 with itself and mate mapped
2494662 + 0 singletons (0.26% : N/A)
11849366 + 0 with mate mapped to a different chr
5389541 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T50.bam`

```
850161108 + 0 in total (QC-passed reads + QC-failed reads)
5104164 + 0 secondary
0 + 0 supplementary
96604362 + 0 duplicates
842887852 + 0 mapped (99.14% : N/A)
845056944 + 0 paired in sequencing
422528472 + 0 read1
422528472 + 0 read2
820050896 + 0 properly paired (97.04% : N/A)
835579702 + 0 with itself and mate mapped
2203986 + 0 singletons (0.26% : N/A)
10011626 + 0 with mate mapped to a different chr
4857008 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T52.bam`

```
824556378 + 0 in total (QC-passed reads + QC-failed reads)
5789060 + 0 secondary
0 + 0 supplementary
86409627 + 0 duplicates
815331372 + 0 mapped (98.88% : N/A)
818767318 + 0 paired in sequencing
409383659 + 0 read1
409383659 + 0 read2
788745460 + 0 properly paired (96.33% : N/A)
807042572 + 0 with itself and mate mapped
2499740 + 0 singletons (0.31% : N/A)
12792368 + 0 with mate mapped to a different chr
7069880 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T55.bam`

```
733508579 + 0 in total (QC-passed reads + QC-failed reads)
3896295 + 0 secondary
0 + 0 supplementary
79263296 + 0 duplicates
729787972 + 0 mapped (99.49% : N/A)
729612284 + 0 paired in sequencing
364806142 + 0 read1
364806142 + 0 read2
712302790 + 0 properly paired (97.63% : N/A)
724168950 + 0 with itself and mate mapped
1722727 + 0 singletons (0.24% : N/A)
7568122 + 0 with mate mapped to a different chr
3447029 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T63.bam`

```
658400145 + 0 in total (QC-passed reads + QC-failed reads)
3903369 + 0 secondary
0 + 0 supplementary
71541242 + 0 duplicates
650984509 + 0 mapped (98.87% : N/A)
654496776 + 0 paired in sequencing
327248388 + 0 read1
327248388 + 0 read2
632700404 + 0 properly paired (96.67% : N/A)
645151286 + 0 with itself and mate mapped
1929854 + 0 singletons (0.29% : N/A)
8101486 + 0 with mate mapped to a different chr
3842350 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T65.bam`

```
752598021 + 0 in total (QC-passed reads + QC-failed reads)
4239687 + 0 secondary
0 + 0 supplementary
83517429 + 0 duplicates
748390865 + 0 mapped (99.44% : N/A)
748358334 + 0 paired in sequencing
374179167 + 0 read1
374179167 + 0 read2
728383450 + 0 properly paired (97.33% : N/A)
742022376 + 0 with itself and mate mapped
2128802 + 0 singletons (0.28% : N/A)
8692440 + 0 with mate mapped to a different chr
4297784 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T69.bam`

```
872922509 + 0 in total (QC-passed reads + QC-failed reads)
5644505 + 0 secondary
0 + 0 supplementary
94807090 + 0 duplicates
867507591 + 0 mapped (99.38% : N/A)
867278004 + 0 paired in sequencing
433639002 + 0 read1
433639002 + 0 read2
840948026 + 0 properly paired (96.96% : N/A)
859019688 + 0 with itself and mate mapped
2843398 + 0 singletons (0.33% : N/A)
12060474 + 0 with mate mapped to a different chr
5779929 + 0 with mate mapped to a different chr (mapQ>=5)
```

`T7.bam`

```
673028951 + 0 in total (QC-passed reads + QC-failed reads)
3919363 + 0 secondary
0 + 0 supplementary
77195580 + 0 duplicates
668220046 + 0 mapped (99.29% : N/A)
669109588 + 0 paired in sequencing
334554794 + 0 read1
334554794 + 0 read2
648922258 + 0 properly paired (96.98% : N/A)
661876850 + 0 with itself and mate mapped
2423833 + 0 singletons (0.36% : N/A)
8514298 + 0 with mate mapped to a different chr
4164353 + 0 with mate mapped to a different chr (mapQ>=5)
```

**Subsample for Ginkgo CNV clustering analysis**  

```
samtools view -q 20 -s 0.04 -b T${NUM}_longranger/phased_possorted_bam.bam | bedtools bamtobed -i stdin > T${NUM}.bed; done
```

[Ginkgo results](http://qb.cshl.edu/ginkgo/?q=results/VPDTv5sfpiEvT9RJ0bRJ)

**CNVnator**

```
cnvnator -root T${NUM}_longranger/out.root -tree T${NUM}_longranger/phased_possorted_bam.bam 
cnvnator -root T${NUM}_longranger/out.root -his 100 -d /work-zfs/mschatz1/resources/grch38/
cnvnator -root T${NUM}_longranger/out.root -stat 100
cnvnator -root T${NUM}_longranger/out.root -partition 100
cnvnator -root T${NUM}_longranger/out.root -call 100 > T${NUM}_longranger/cnvnator.tsv'
```

Base pairs affected by cnvnator +5bp buffer
```
103
821081155
106
449644488
110
1279156262
16
425740413
23
501690372
28
572813326
30
903679837
33
821044269
35
373917211
43R
1276597117
49
816192070
50
885429250
51
1028786189
54
468488551
55
392017013
57
1058318633
60
1081253604
63
459658809
65
1146712512
70
589217438
```

## 1. Train

```
for NUM in 103 106 110 16 23 28 30 33 35 43R 49 50 51 54 55 57 60 63 65 70; do 

gunzip -c T${NUM}_longranger/phased_variants.vcf.gz > T${NUM}_longranger/phased_variants.vcf; done

python /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/0.setup/generateVarfile.py --vcf T${NUM}_longranger/phased_variants.vcf --out T${NUM}_longranger/out.varfile --fai /work-zfs/mschatz1/resources/refdata-GRCh38-2.1.0/fasta/genome.fa.fai --variantspacing 80000; done

python /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/1.simulate/simulate.py --simulate --varfile T${NUM}_longranger/out.varfile --nproc 16 --bam T${NUM}_longranger/phased_possorted_bam.bam > T${NUM}_longranger/mosaic.features.tsv; done

shuf -o T${NUM}_longranger/phased_variants.vcf T${NUM}_longranger/phased_variants.vcf; done

sbatch --job-name=HET${NUM} --partition=shared --nodes=1 --ntasks-per-node=12 --export=ALL --time=4:0:0 --workdir=/scratch/groups/mschatz1/cdarby/brca/ --wrap="python /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/1.simulate/simulate.py --het --max 50000 --varfile T${NUM}_longranger/phased_variants.vcf --nproc 11 --bam T${NUM}_longranger/phased_possorted_bam.bam > T${NUM}_longranger/het.features.tsv"

sbatch --job-name=HOM${NUM} --partition=shared --nodes=1 --ntasks-per-node=12 --export=ALL --time=4:0:0 --workdir=/scratch/groups/mschatz1/cdarby/brca/ --wrap="python /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/1.simulate/simulate.py --hom --max 50000 --varfile T${NUM}_longranger/phased_variants.vcf --nproc 11 --bam T${NUM}_longranger/phased_possorted_bam.bam > T${NUM}_longranger/hom.features.tsv"

python /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/2.train/train.py --hom T${NUM}_longranger/hom.features.tsv --het T${NUM}_longranger/het.features.tsv --mosaic T${NUM}_longranger/mosaic.features.tsv --out T${NUM}_longranger/clf.pkl; done

```

Number of mosaic training examples (equal number germline, of which equally divided between hom and het) for each sample, aimed for ~20K.

103
23450

106
24951

110
21576

16
20170

23
24035

28
25183

30
19372

33
21512

35
23578

43R
17303

49
20854

50
22027

51
23357

54
20306

55
25566

57
21643

60
18255

63
20000

65
21397

70
22426

## 2. Scan

```
sbatch --job-name=SCAN${NUM} --partition=shared --nodes=1 --ntasks-per-node=12 --export=ALL --time=48:0:0 --workdir=/scratch/groups/mschatz1/cdarby/brca/ --wrap="python /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/3.classify/classify.py --nproc 11 --bam T${NUM}_longranger/phased_possorted_bam.bam --clf T${NUM}_longranger/clf.pkl --bed scripts/windows.tsv  > T${NUM}_longranger/scan.tsv" ; done
```

Filter on repeat regions 

```
bedtools intersect -v -a T${NUM}_longranger/scan.tsv -b ~/scratch/repeatasm/src/repeatsGRCh38.bed > T${NUM}_longranger/scan_norepeat.bed; done
```

Compute features  

```
python /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/4.filter/computeFilterFeatures.py --bam T${NUM}_longranger/phased_possorted_bam.bam --bed T${NUM}_longranger/scan_norepeat.bed --nproc X > T${NUM}_longranger/filter.features.tsv & done
```

Filter on features 

```
Rscript /work-zfs/mschatz1/cdarby/HCC1954T/pipeline/4.filter/filter.R T${NUM}_longranger/filter.features.tsv T${NUM}_longranger/calls.tsv; done
``` 

Compare calls against bamsurgeon mutations & validation set  

```
join -t $'\t' -j1 ../variants0idx.txt scanNB_norepeat.features.tsv | awk '{print "2\t" $0}' > scanNB_norepeat_labeled.features.tsv
join -t $'\t' -j1 ../sitesN.txt scanNB_norepeat.features.tsv | awk '{print "1\t" $0}' >> scanNB_norepeat_labeled.features.tsv
cat scanNB_norepeat.features.tsv | awk '{print "0\t" $0}' >> scanNB_norepeat_labeled.features.tsv
sort -u -k2 -o scanNB_norepeat_labeled.features.tsv scanNB_norepeat_labeled.features.tsv
```

```
    1294 T103_longranger/calls1.tsv
     755 T106_longranger/calls1.tsv
    1060 T110_longranger/calls1.tsv
     231 T16_longranger/calls1.tsv
     654 T23_longranger/calls1.tsv
    1003 T28_longranger/calls1.tsv
    1305 T30_longranger/calls1.tsv
     640 T33_longranger/calls1.tsv
     499 T35_longranger/calls1.tsv
     474 T43R_longranger/calls1.tsv
     267 T49_longranger/calls1.tsv
    1784 T50_longranger/calls1.tsv
    1319 T51_longranger/calls1.tsv
     367 T54_longranger/calls1.tsv
     931 T55_longranger/calls1.tsv
     929 T57_longranger/calls1.tsv
     452 T60_longranger/calls1.tsv
     744 T63_longranger/calls1.tsv
     982 T65_longranger/calls1.tsv
     465 T70_longranger/calls1.tsv
```

##3. MosaicHunter 


Single: (20) 10X; (14) Illumina
Paired: (6) 10X Tumor / Illumina blood & tumor ; (14) Illumina Tumor / Illumina blood
35, 43R, 50, 55, 63, 65 have both Illumina T/B + 10X T

##4. HapMuc





















